README.md
# minForth

The main goals for this minimal Forth are as follows:

- To be very frugal with its usage of memory.
- To be easy to modify and add/extend the primitives.
- To have an implementation that is minimal and "intuitively obvious upon casual inspection".
- To be able to run on any system that has a C compiler.
- To be deployable to as many different kinds of development boards as possible via the Arduino IDE.
- To have a VML (virtual machine language) that is as human-reabable as possible, but not at the expense of performance.

To these ends, I have wandered off the beaten path in the following ways:

- This is NOT an ANSI-standard Forth.
- This is a byte-coded implementation to save code space.
- Many primitives (core words) are built into the compiler, and are not included in the dictionary.
- The primitves ARE NOT case sensitive (DUP = dup = Dup).
- User-defined words ARE case sensitive.
- The dictionary is intersparsed with the CODE; it is not separated.
- A dictionary entry looks like this: (offset,flags,name,null terminator,implementation).
- It has only 3 additional bytes of overhead (in addition to the name).
- To save space, code addresses are 2 bytes, so code space is limited to 16 bits (64kb).
- All CODE addresses are offsets into the CODE space, not absolute addresses.
- HERE and LAST are also offsets into the CODE space, not absolute addresses.
- The VARIABLE space is separated from the CODE space, and can be larger than 64kb.
- VHERE is a 32-bit absolute address to the first available byte in VARIABLE.
- There are 10 temporary words (T0..T9) that can be re-defined without any dictionary overhead.
- There are 10 temporary variables (r0..r9) that can be allocated/destroyed.
- "WORDS" is built in, and cannot be called by another word at runtime (it's IMMEDIATE).

## Temporary words:
- A temporary word is named T0 .. T9. (e.g. - ": T4 ... ;" will define word T4, but T4 is not added to the dictionary)
- I think of temporary words as named transient words that don't incur any dictionary overhead.
- Their purpose is to support code reuse and factoring without the overhead of a dictionary entriy.
- You cannot create an IMMEDIATE temporary word.
- When defined, they refer to the current value of HERE.
- All previous references to T4 still refer to the previous definition. New references to T4 refer to the new definition.
- Here is a simple example:
```
: T1 dup $20 < if drop '.' then ;
: dumpC for i c@ T1 emit next ;
: .code cb dup here + 1- for i c@ T1 emit next ;
```

## Temporary variables:
- Temporary variables are allocated in sets of 10.
- They are completely under the control of the programmer.
- They are not built into the call sequence, so they can be accessed across words.
- They are not set or retrieved using @ and !.
- There are 2 special operations for temp variables: read and set (eg - r3 1+ s3).
- The words to manage temps are: +tmps, r0..r9, s0..s9, i0..i9, d0..d9 and -tmps
  - +tmps: allocate 10 new temps
  - rX: push the value of temp #x onto the stack (read)
  - sX: pop the new value for temp #X off the stack (set)
  - iX: increment temp #X
  - dX: decrement temp #x
  - -tmps: destroy the most recently allocated temps
- Here are some simple examples:
```
: betw +tmps s3 s2 s1 r2 r1 <= r1 r3 <= and -tmps ;
: .c (n--) s9 r9 #32 $7e betw if r9 emit else r9 ." (%d)" then ;
: dumpX (a n--) swap s8 1 for r8 c@ .c i8 next ;
```


history.txt
bye
bye
ex
exy
bye
0 edit
0 edit


bye
: tt ( a b -- c ) 1 2 3 + + . ;
tt
words
bye
words
: ttt 1 2 3 + + . 10 100 for i . next ;
ttt
bye
: rr 100 begin dup . 1- while ;
rr
bye
bye
bye
bye
bye
words
r1 .
r0 .
223 s1
r1 .
lt
r1 .
73466 s1
r1
.
lt
r1 .
82773 s1
r1 .
lt
lt
r1 .
38747 s1
r1 .
lt
r1 .
bye
bye
bye
bye
1 8 >>
1 8 <<
drop
drop
256 8 >>
1 <<
1 <<
1 <<
1 >>
1 >>
1 >>
1 >>
bye
500 mil
words
bm
900 mil bm
100 mil bm
: dd 1 for i mil bm cr next ;
3 dd
10 dd
500 600 dd
: dd 1 for i mil 100 * bm cr next ;
10 dd
: dd 1 for i mil 100 * dup . bm cr next ;
10 dd
: dd 1 for cr i mil 100 * bm next ;
10 dd
: dd 1 for cr i mil 100 * dup . bm next ;
10 dd
words
h @ .
reset
words
bye
: dd 1 for cr i mil 100 * dup . bm next ;
5 bm
5 dd
5 dd
5 dd
5 dd
5 dd
: sd 1 10 for dd next ;
sd
: sd 1 10 for 5 dd next ;
sd
words
: here h @ ;
here .
u
.
: here h @ u + ;
here u - .
: .c dup 32 > if drop '.' then emit ;
: dump for i c@ emit next ;
: dump for i c@ .c next ;
words
u here dump
: .c dup 32 < if drop '.' then emit ;
: dump for i c@ .c next ;
u here dump
here u - .
words
bye
bye
bye



words
1 load
words
1 load
1
load
words
1 edit
1 edit
bye
.code
here .
bye
words
words
500 mil
bm
bye
words
reset
words
0 load
words
bye
" hi there"
bye
" this is %d"
: test " hi there" ;
;
test
bye
: ttt " hi there" ;
ttt
.cod
.code
" hi there 123"
223
" %d"
.

" %d"
223 " %h"
223 " %x"
" %d"
123 dup " %d %x"
" test"
223 dup dup " %d %x %b"
bye
." tedt"
" ttee"
bye
" test"
type
." test2"
" test"
typ
type
." TEST"
." hello world"
" test" type
bye
1 2 3 " test %d %b" .
." test %d %b" .
: tt1 " this is a test %q%x%q" type ;
4475 tt1
$3877
tt1
bye
words
.code
ha
hex
dup dup ." %d %x %b"
bye
: lt s2 for i r2 = if ." brk" break then i . next ." %bend." ;
100 200 133 lt
bye
." test" ."  for ecc"
" test " "  12345"
type type
"  test%q" " %q12345" type type
: .ch s9 r9 r9 r9 r9 ." %d %x %x %c%n" ;
44 .ch
: .ch s9 r9 r9 r9 r9 ." %d (%c) %x %b%n" ;
44 .ch
: ascii 32 126 for i .ch next ;
ascii
r9 .
$20 $7e
. .
words
.code
last
words
last
user
+
c@
.code
words
38 emit
123
.code
.
3 4 5
s3 s2 s1
r1 r2 < 0=
r1 r3 > 0=
r3 .
r1 .
r1 r3
>
0=
. . 
. 
4 3 5
s3 s2 s1
r1 r2 < 0=
r1 r3 > 0=
=
.
0 load
+tmps
-tmps
.
0 load
words
.code
223
,code
.code
44
T1
33 44 T1
T1
31
T1
44 22 T1
T1
0 load
.code
13445
.code
.
here .
last .
u
user
.
last
.
la
@
user
+
c@
last
swap
-
dup
user +
c@
.
la 
!
words
.code
: test 12 34 ;
test
words
. .
forget
words
test
here
test
: test 12 34 ; test . . 
here
forget
here
. .
. 
last here
: test 12 34 ; test . . 
last here
forget
last here
. . . . . . . .
words
.code
44 33 88 betw .
22 33 88 betw .
22 11 15 betw .
22 11 22 betw .
22 22 22 betw .
22 21 23 betw .
25 21 23 betw .
20 21 23 betw .
.
bye
words
.code
here .
34 T1
22 T1
: dd 100 200 for i i ." : w%d .%q w%d %q ;" next ;
dd
forget
: dd 100 200 for i i ." : w%d .%q w%d %q ;%n" next ;
dd
forget
: dd 100 200 for i i ." : w%d .%q -w%d-%q ;%n" next ;
dd
: w176 ." w176 " ;
: w177 ." w177 " ;
: w178 ." w178 " ;
: w179 ." w179 " ;
: w180 ." w180 " ;
: w181 ." w181 " ;
: w182 ." w182 " ;
: w183 ." w183 " ;
: w184 ." w184 " ;
: w185 ." w185 " ;
: w186 ." w186 " ;
: w187 ." w187 " ;
: w188 ." w188 " ;
: w189 ." w189 " ;
: w190 ." w190 " ;
: w191 ." w191 " ;
: w192 ." w192 " ;
: w193 ." w193 " ;
: w194 ." w194 " ;
: w195 ." w195 " ;
: w196 ." w196 " ;
: w197 ." w197 " ;
: w198 ." w198 " ;
: w199 ." w199 " ;
: w200 ." w200 " ;

OK ()>forget

OK ()>: dd 100 200 for i i ." : w%d .%q -w%d-%q ;%n" next ;

OK ()>dd
: w100 ." -w100-" ;
: w101 ." -w101-" ;
: w102 ." -w102-" ;
: w103 ." -w103-" ;
: w104 ." -w104-" ;
: w105 ." -w105-" ;
: w106 ." -w106-" ;
: w107 ." -w107-" ;
: w108 ." -w108-" ;
: w109 ." -w109-" ;
: w110 ." -w110-" ;
: w111 ." -w111-" ;
: w112 ." -w112-" ;
: w113 ." -w113-" ;
: w114 ." -w114-" ;
: w115 ." -w115-" ;
: w116 ." -w116-" ;
: w117 ." -w117-" ;
: w118 ." -w118-" ;
: w119 ." -w119-" ;
: w120 ." -w120-" ;
: w121 ." -w121-" ;
: w122 ." -w122-" ;
: w123 ." -w123-" ;
: w124 ." -w124-" ;
: w125 ." -w125-" ;
: w126 ." -w126-" ;
: w127 ." -w127-" ;
: w128 ." -w128-" ;
: w129 ." -w129-" ;
: w130 ." -w130-" ;
: w131 ." -w131-" ;
: w132 ." -w132-" ;
: w133 ." -w133-" ;
: w134 ." -w134-" ;
: w135 ." -w135-" ;
: w136 ." -w136-" ;
: w137 ." -w137-" ;
: w138 ." -w138-" ;
: w139 ." -w139-" ;
: w140 ." -w140-" ;
: w141 ." -w141-" ;
: w142 ." -w142-" ;
: w143 ." -w143-" ;
: w144 ." -w144-" ;
: w145 ." -w145-" ;
: w146 ." -w146-" ;
: w147 ." -w147-" ;
: w148 ." -w148-" ;
: w149 ." -w149-" ;
: w150 ." -w150-" ;
: w151 ." -w151-" ;
: w152 ." -w152-" ;
: w153 ." -w153-" ;
: w154 ." -w154-" ;
: w155 ." -w155-" ;
: w156 ." -w156-" ;
: w157 ." -w157-" ;
: w158 ." -w158-" ;
: w159 ." -w159-" ;
: w160 ." -w160-" ;
: w161 ." -w161-" ;
: w162 ." -w162-" ;
: w163 ." -w163-" ;
: w164 ." -w164-" ;
: w165 ." -w165-" ;
: w166 ." -w166-" ;
: w167 ." -w167-" ;
: w168 ." -w168-" ;
: w169 ." -w169-" ;
: w170 ." -w170-" ;
: w171 ." -w171-" ;
: w172 ." -w172-" ;
: w173 ." -w173-" ;
: w174 ." -w174-" ;
: w175 ." -w175-" ;
: w176 ." -w176-" ;
: w177 ." -w177-" ;
: w178 ." -w178-" ;
: w179 ." -w179-" ;
: w180 ." -w180-" ;
: w181 ." -w181-" ;
: w182 ." -w182-" ;
: w183 ." -w183-" ;
: w184 ." -w184-" ;
: w185 ." -w185-" ;
: w186 ." -w186-" ;
: w187 ." -w187-" ;
: w188 ." -w188-" ;
: w189 ." -w189-" ;
: w190 ." -w190-" ;
: w191 ." -w191-" ;
: w192 ." -w192-" ;
: w193 ." -w193-" ;
: w194 ." -w194-" ;
: w195 ." -w195-" ;
: w196 ." -w196-" ;
: w197 ." -w197-" ;
: w198 ." -w198-" ;
: w199 ." -w199-" ;
: w200 ." -w200-" ;
words
w121
.
w200

minForth.cpp
// MinForth.cpp : An extremely memory conscious Forth interpreter

#include "Shared.h"

typedef struct {
    const char *name;
    const char *op;
} PRIM_T;

// Words that directly map to VM operations
PRIM_T prims[] = {
    {"+","+"}, {"-","-"}, {"/","/"}, {"*","*"},{"/mod","&"},
    {"swap","$"}, {"drop","\\"}, {"over","%"}, {"dup","#"},
    {"nip","$\\"}, {"tuck","$%"}, {"2dup","%%"}, {"2drop","\\\\"},
    {"emit",","}, {"(.)","."}, {"space","b"}, {"cr","n"}, {"bl","1 "},
    {"=","="}, {"<","<"}, {">",">"}, 
    {"<=",">N"}, {"<>","=N"}, {"!=","=N"}, {">=","<N"}, {"0=","N"},
    {"<<","L"}, {">>","R"}, {"zlen","T"}, {".",".b"},
    {"@","@"}, {"c@","c"}, {"w@","w"}, {"!","!"}, {"c!","C"}, {"w!","W"},
    {"and","a"}, {"or","o"}, {"xor","x"}, {"com","~"}, {"not","N"},
    {"1+","i"}, {"1-","d"}, {"I", "I"}, {"+I", "m"}, {"execute","G"},
    {"leave",";"}, {"timer","zT"}, {"wait","zW"}, {"reset","Y"}, {"break","^"},
    {"key","K"}, {"key?","?"}, 
    {"+tmps","p"}, {"-tmps","q"},
    {"ztype","Z"}, {"qtype","t"},
    // Extensions
#if __BOARD__ == PC
    { "bye","zZ" }, {"load","zL"},
#else
    { "p-i","zPO" }, { "p-o","zPI" }, { "p-u","zPU" },
    { "a-r","zAR" }, { "a-w","zAW" }, { "d-r","zDR" }, { "d-w","zDW" },
#endif
#ifdef __EDITOR__
    { "edit","zE" },
#endif
#ifdef __GAMEPAD__
    // Extensions
    { "gp-button","xGB" },
#endif
    {"nop"," "}, {0,0}
};

char word[32], *in;
byte lastWasCall = 0, isBye = 0;
byte *VHERE, *VHERE2;
CELL HERE, LAST, STATE, tempWords[10];

void CComma(CELL v) { code[HERE++] = (byte)v; }
void Comma(CELL v) { SET_LONG(&code[HERE], v); HERE += CELL_SZ; }
void WComma(WORD v) { SET_WORD(&code[HERE], v); HERE += 2; }

char lower(char c) { return betw(c, 'A', 'Z') ? (c + 32) : c; }

byte strEq(const char* x, const char* y) {
    while (*x && *y && (*x == *y)) { ++x; ++y; }
    return (*x || *y) ? 0 : 1;
}

byte strEqI(const char* x, const char* y) {
    while (*x && *y) {
        if (lower(*x) != lower(*y)) { return 0; }
        ++x; ++y;
    }
    return (*x || *y) ? 0 : 1;
}

char *strCpy(char* d, const char* s) {
    while (*s) { *(d++) = *(s++); }
    *d = 0;
    return d;
}

int strLen(const char* str) {
    int l = 0;;
    while (*(str++)) { ++l; }
    return l;
}

void printStringF(const char* fmt, ...) {
    char* buf = (char*)&code[CODE_SZ-64];
    va_list args;
    va_start(args, fmt);
    vsnprintf(buf, 100, fmt, args);
    va_end(args);
    printString(buf);
}

int isTempWord(const char* nm) {
    return ((nm[0] == 'T') && betw(nm[1], '0', '9') && (nm[2] == 0));
}

void doCreate(const char* name, byte f) {
    if (isTempWord(name)) {
        tempWords[name[1] - '0'] = HERE;
        return;
    }
    DICT_T *dp = DP_AT(HERE);
    dp->prev = (byte)(HERE - LAST);
    dp->flags = f;
    strCpy(dp->name, name);
    LAST = HERE;
    HERE += strLen(name) + 3;
}

int doFind(const char* name) {
    // Temporary word?
    if (isTempWord(name) && (tempWords[name[1] - '0'])) {
        push(tempWords[name[1] - '0']);
        push(0);
        return 1;
    }
    CELL def = (WORD)LAST;
    while (def) {
        DICT_T* dp = DP_AT(def);
        if (strEq(dp->name, name)) {
            push(def + strLen(dp->name) + 3);
            push(dp->flags);
            return 1;
        }
        if (def == dp->prev) break;
        def -= dp->prev;
    }
    return 0;
}

int doWords() {
    CELL l = (WORD)LAST;
    while (l) {
        DICT_T *dp = DP_AT(l);
        printString(dp->name);
        printChar(' ');
        if (l == dp->prev) break;
        l -= dp->prev;
    }
    return 1;
}

int getWord(char* wd) {
    while (*in && (*in < 33)) { ++in; }
    int l = 0;
    while (*in && (32 < *in)) {
        *(wd++) = *(in++);
        ++l;
    }
    *wd = 0;
    return l;
}

int execWord() {
    CELL f = pop(), xt = pop();
    if ((STATE) && (f == 0)) {
        CComma(':');
        WComma((WORD)xt);
        lastWasCall = 1;
    } else { run((WORD)xt); }
    return 1;
}

int doNumber(int state) {
    if (state) {
        CELL num = pop();
        if ((num & 0xFF) == num) {
            CComma('1');
            CComma(num & 0xff);
        }
        else if ((num & 0xFFFF) == num) {
            CComma('2');
            WComma((WORD)num);
        }
        else {
            CComma('4');
            Comma(num);
        }
    }
    return 1;
}

int isNum(const char* wd) {
    CELL x = 0;
    int base = BASE, isNeg = 0, lastCh = '9';
    if ((wd[0]=='\'') && (wd[2]==wd[0]) && (wd[3]==0)) { push(wd[1]); return 1; }
    if (*wd == '#') { base = 10;  ++wd; }
    if (*wd == '$') { base = 16;  ++wd; }
    if (*wd == '%') { base = 2;  ++wd; lastCh = '1'; }
    if ((*wd == '-') && (base == 10)) { isNeg = 1;  ++wd; }
    if (*wd == 0) { return 0; }
    while (*wd) {
        char c = *(wd++);
        int t = -1;
        if (betw(c, '0', lastCh)) { t = c - '0'; }
        if ((base == 16) && (betw(c, 'A', 'F'))) { t = c - 'A' + 10; }
        if ((base == 16) && (betw(c, 'a', 'f'))) { t = c - 'a' + 10; }
        if (t < 0) { return 0; }
        x = (x * base) + t;
    }
    if (isNeg) { x = -x; }
    push(x);
    return 1;
}

int doPrim(const char* wd) {
    const char* vml = NULL;
    char x[3];
    if (betw(wd[0],'r','s') && betw(wd[1],'0','9') && (!wd[2])) { vml = wd; }
    if ((wd[0]=='i') && betw(wd[1], '0', '9') && (!wd[2])) { strCpy(x, wd); x[0] = 'e'; vml = x; }
    if ((wd[0]=='d') && betw(wd[1], '0', '9') && (!wd[2])) { strCpy(x, wd); x[0] = 'f'; vml = x; }
    for (int i = 0; prims[i].op && (!vml); i++) {
        if (strEqI(prims[i].name, wd)) { vml = prims[i].op; }
    }
    if (!vml) { return 0; }
    if (STATE) {
        for (int j = 0; vml[j]; j++) { CComma(vml[j]); }
    } else {
        byte* x = CA(HERE + 10);
        for (int j = 0; vml[j]; j++) { *(x++) = vml[j]; }
        *x = ';';
        run((WORD)HERE + 10);
    }
    return 1;
}

int doQuote() {
    in++;
    if (STATE) { VHERE2 = VHERE; }
    push((CELL)VHERE2);
    while (*in && (*in != '"')) { *(VHERE2++) = *(in++); }
    *(VHERE2++) = 0;
    if (*in) { ++in; }
    if (STATE) {
        CComma('4');
        Comma(pop());
        VHERE = VHERE2;
    }
    return 1;
}

int doDotQuote() {
    doQuote();
    if (STATE) {
        CComma('Z');
    } else {
        byte* cp = CA(HERE+10);
        *(cp++) = 'Z';
        *(cp++) = ';';
        run((WORD)(HERE+10));
    }
    return 1;
}

int doParseWord(char* wd) {
    byte lwc = lastWasCall;
    lastWasCall = 0;

    if (doPrim(wd)) { return 1; }
    if (doFind(wd)) { return execWord(); }
    if (isNum(wd)) { return doNumber(STATE); }
    if (strEq(wd, ".\"")) { return doDotQuote(); }
    if (strEq(wd, "\"")) { return doQuote(); }

    if (strEq(wd, ":")) {
        if (getWord(wd)) {
            doCreate(wd, 0);
            STATE = 1;
        }
        else { return 0; }
        return 1;
    }

    if (strEq(wd, "(")) {
        while (*in && (*in != ')')) { ++in; }
        if (*in == ')') { ++in; }
        return 1;
    }

    if (strEq(wd, ";")) {
        STATE = 0;
        if (lwc && (code[HERE - 3] == ':')) { code[HERE - 3] = 'J'; }
        else { CComma(';'); }
        return 1;
    }

    if (strEqI(wd, "VARIABLE")) {
        push((CELL)VHERE);
        VHERE += CELL_SZ;
        strCpy(wd, "CONSTANT");
    }

    if (strEqI(wd, "CONSTANT")) {
        if (getWord(wd)) {
            doCreate(wd, 0);
            doNumber(1);
            CComma(';');
            return 1;
        }
        else { return 0; }
    }

    if (strEqI(wd, "IMMEDIATE")) { DP_AT(LAST)->flags |= 1; return 1; }
    if (strEqI(wd, "ALLOT")) { VHERE += pop();              return 1; }
    if (strEqI(wd, "WORDS")) { return doWords(); }

    if (strEqI(wd, "IF")) {
        CComma('j');
        push(HERE);
        WComma(0);
        return 1;
    }

    if (strEqI(wd, "ELSE")) {
        CELL tgt = pop();
        CComma('J');
        push(HERE);
        WComma(0);
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "THEN")) {
        CELL tgt = pop();
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "FOR")) {
        CComma('[');
        push(HERE);
        WComma(0);
        return 1;
    }

    if (strEqI(wd, "NEXT")) {
        CComma(']');
        CELL tgt = pop();
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "BEGIN")) {
        CComma('{');
        push(HERE);
        WComma(0);
        return 1;
    }

    if (strEqI(wd, "AGAIN")) {
        CComma('}');
        CELL tgt = pop();
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "WHILE")) {
        CComma('v');
        CELL tgt = pop();
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "UNTIL")) {
        CComma('u');
        CELL tgt = pop();
        SET_WORD(CA(tgt), (WORD)HERE);
        return 1;
    }

    if (strEqI(wd, "FORGET")) {
        HERE = LAST;
        LAST -= CODE(LAST);
        return 1;
    }

    STATE = 0;
    VHERE2 = VHERE;
    printStringF("[%s]??", wd);
    return 0;
}

void doParse(const char* line) {
    in = (char*)line;
    int len = getWord(word);
    while (0 < len) {
        if (VHERE2 < VHERE) { VHERE2 = VHERE; }
        if (strEq(word, "//")) { return; }
        if (strEq(word, "\\")) { return; }
        if (doParseWord(word) == 0) { return; }
        len = getWord(word);
    }
}

void doOK() {
    if (STATE) { printString(" ... "); return; }
    printString("\r\nOK (");
    for (int d = 1; d <= sp; d++) {
        if (1 < d) { printChar(' '); }
        printBase(stk[d], BASE);
    }
    printString(")>");
}

char *rtrim(char* str) {
    char* cp = str;
    while (*cp) { ++cp; }
    --cp;
    while ((str <= cp) && (*cp <= ' ')) { *(cp--) = 0; }
    return str;
}

void systemWords() {
    char* cp = (char*)(VHERE + 6);
    sprintf(cp, ": cb %lu ;", (UCELL)code);     doParse(cp);
    sprintf(cp, ": vb %lu ;", (UCELL)vars);     doParse(cp);
    sprintf(cp, ": csz %d ;", CODE_SZ);         doParse(cp);
    sprintf(cp, ": vsz %d ;", VARS_SZ);         doParse(cp);
    sprintf(cp, ": ha %lu ;", (UCELL)&HERE);    doParse(cp);
    sprintf(cp, ": la %lu ;", (UCELL)&LAST);    doParse(cp);
    sprintf(cp, ": va %lu ;", (UCELL)&VHERE);   doParse(cp);
    sprintf(cp, ": base %lu ;", (UCELL)&BASE);  doParse(cp);
}

#if __BOARD__ == PC
FILE* input_fp = NULL;
FILE* fpStk[10];
byte fpSP = 0;

void fpPush(FILE* v) { if (fpSP < 9) { fpStk[++fpSP] = v; } }
FILE* fpPop() { return (fpSP) ? fpStk[fpSP--] : 0 ; }

void doLoad(int blk) {
    char* fn = (char*)(VHERE + 100);
    sprintf(fn, "./block-%03d.4th", blk);
    FILE* fp = fopen(fn, "rt");
    if (fp) {
        if (input_fp) { fpPush(input_fp); }
        input_fp = fp;
    }
}

byte *doExt(CELL ir, byte *pc) {
    switch (ir) {
    case 'E': doEditor();                       break;
    case 'L': doLoad(pop());                    break;
    case 'T': push(GetTickCount());             break;
    case 'W': if (TOS) { Sleep(TOS); } pop();   break;
    case 'Z': isBye = 1;                        break;
    default: printString("-unk ext-");
    }
    return pc;
}

void printString(const char* cp) { printf("%s", cp); }
void printChar(char c) { printf("%c", c); }
int charAvailable(void) { return _kbhit(); }
int getChar(void) { return _getch(); }


void doHistory(const char* txt) {
    FILE* fp = fopen("history.txt", "at");
    if (fp) {
        fputs(txt, fp);
        fclose(fp);
    }
}

void loop() {
    char *tib = (char *)CA(HERE+32);
    FILE* fp = (input_fp) ? input_fp : stdin;
    if (fp == stdin) { doOK(); }
    if (fgets(tib, 100, fp) == tib) {
        if (fp == stdin) { doHistory(tib); }
        doParse(rtrim(tib));
        return;
    }
    if (input_fp) {
        fclose(input_fp);
        input_fp = fpPop();
    }
}

int main()
{
    vmReset();
    doLoad(0);
    while (!isBye) { loop(); }
}

#endif
